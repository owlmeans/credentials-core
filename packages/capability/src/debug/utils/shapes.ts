
import {
  credentialShape,
  didShape,
  doubleDidShape,
  proofShape
} from '@owlmeans/regov-ssi-agent/src/debug/utils/shapes'


export const organizationDid = {
  ...(did => {
    const _did = {...did};
    // delete (<any>_did).authentication;

    ((<any>_did).verificationMethod as Array<unknown>).splice(0, 1)
    
    return _did
  })(doubleDidShape),
  capabilityInvocation: [
    expect.any(String)
  ]
}

export const capabilityDid = {
  ...(did => {
    const _did = {...did}
    delete (<any>_did).authentication
    
    return _did
  })(doubleDidShape),
  capabilityInvocation: [
    expect.any(String)
  ]
}

export const capaiblityDoubleDid = {
  assertionMethod: [
    expect.any(String),
    expect.any(String),
  ],
  authentication: [
    expect.any(String),
  ],
  capabilityInvocation: [
    expect.any(String),
  ],
  id: expect.any(String),
  verificationMethod: [
    {
      controller: expect.any(String),
      id: expect.any(String),
      nonce: expect.any(String),
      publicKeyBase58: expect.any(String),
    },
    {
      controller: expect.any(String),
      id: expect.any(String),
      nonce: expect.any(String),
      publicKeyBase58: expect.any(String),
    }
  ],
  proof: proofShape
}

export const entityShape = {
  ...credentialShape,
  credentialSubject: {
    data: {
      identity: credentialShape
    },
    did: didShape
  }
}

export const orgCapaiblityShape = {
  ...credentialShape,
  credentialSubject: {
    source: {
      ...credentialShape,
      credentialSubject: {
        source: credentialShape,
        sourceDid: didShape
      }
    },
    sourceDid: organizationDid
  }
}

export const membershipCapabilityShape = {
  ...credentialShape,
  credentialSubject: {
    data: {
      defaults: {
        organziationDid: expect.any(String)
      }
    },
    source: orgCapaiblityShape,
    sourceDid: capaiblityDoubleDid
  }
}