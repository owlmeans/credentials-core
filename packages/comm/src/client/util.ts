/**
 *  Copyright 2023 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import {
  CredentialWrapper, Credential, ExtensionRegistry, EXTENSION_TRIGGER_INCOMMING_DOC_RECEIVED, RegistryItem,
  WalletWrapper, Presentation
} from "@owlmeans/vc-core"
import { IncommigDocumentWithConn, IncommingWrapperMeta } from "../types"


export const triggerIncommingDocView = async <
  Subject extends {} = {},
  Type extends RegistryItem<Subject> = Presentation<Credential<Subject>>,
  Meta extends IncommingWrapperMeta = IncommingWrapperMeta
>(
  registry: ExtensionRegistry, wallet: WalletWrapper, wrapper: CredentialWrapper<Subject, Type, Meta>
) => {
  await registry.triggerEvent<IncommigDocumentWithConn>(
    wallet, EXTENSION_TRIGGER_INCOMMING_DOC_RECEIVED, {
    credential: wrapper.credential as unknown as Presentation, statusHandler: { successful: false },
    conn: (wrapper.meta as IncommingWrapperMeta).conn,
    cleanUp: () => undefined
  })
}