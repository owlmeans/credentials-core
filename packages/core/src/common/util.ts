/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { MaybeArray, Idish } from "./types"

export const normalizeValue = <Type>(value: MaybeArray<Type>): Type[] => {
  return Array.isArray(value)
    ? value
    : value ? [value] : []
}

export const simplifyValue = <Type>(value: MaybeArray<Type>) => {
  return Array.isArray(value) && value.length < 2
    ? value.length === 0 ? undefined : value[0]
    : value
}

export const singleValue = <Type>(value: MaybeArray<Type>): Type | undefined => {
  return Array.isArray(value)
    ? value.length > 0 ? value[0] : undefined
    : value
}

export const mapValue = async <Type, Res>(
  value: MaybeArray<Type>,
  callback: (value: Type) => Promise<Res>
): Promise<Res[]> => {
  return Promise.all(normalizeValue(value).map(callback))
}

export const addToValue = <Type>(value: MaybeArray<Type> | undefined, add: MaybeArray<Type>): MaybeArray<Type> => {
  if (value) {
    return [...normalizeValue(value), ...normalizeValue(add)]
  }

  return add
}

export const extractId = (id: Idish) => typeof id === 'string' ? id : id.id

export const getDeepValue = <Type, Subject extends {}>(subject: Subject, key: string, def?: Type): Type => {
  const parts = key.split('.')
  return parts.reduce((value, key) => {
    if (subject.hasOwnProperty(key)) {
      const tmp = (subject as any)[key]
      subject = tmp

      return tmp
    }

    return def
  }, def) as Type
}