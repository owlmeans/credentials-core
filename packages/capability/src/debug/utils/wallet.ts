import { nodeCryptoHelper } from "@owlmeans/regov-ssi-common"
import {
  buildWalletWrapper,
  Presentation,
  REGISTRY_TYPE_REQUESTS,
  Credential,
  CredentialSubject,
  WrappedDocument,
  UnsignedCredential,
  REGISTRY_TYPE_IDENTITIES,
  Identity
} from "@owlmeans/regov-ssi-core"
import {
  EntityIdentity,
  identityHelper,
  issuerCredentialHelper,
  holderCredentialHelper,
  verifierCredentialHelper,
  RequestBundle,
  ClaimCredential,
  ClaimSubject,
  OfferCredential,
  OfferSubject,
  OfferBundle,
  ClaimBundle,
  HolderVisitor,
  CREDENTIAL_ENTITY_IDENTITY_TYPE
} from "@owlmeans/regov-ssi-agent"
import { governanceCredentialHelper } from "../../governance/credential"
import {
  CapabilityCredential,
  CapabilityDocument,
  CapabilityExtension,
  ClaimCapability,
  CREDENTIAL_GOVERNANCE_TYPE,
  OfferCapability,
  OfferCapabilityExtension,
  REGISTRY_TYPE_CAPABILITY,
  UnsignedCapabilityCredential
} from "../../governance/types"
import { holderGovernanceVisitor } from '../../governance/holder'

import { TestUtil as AgentTestUtil } from '@owlmeans/regov-ssi-agent/src/debug/utils/wallet'
import { verifierCapabilityHelper } from "../../verifier/capability"
import { issuerVisitor } from "../../issuer/capability"
import { ByCapabilityExtension } from "../../issuer/types"
import { holderCapabilityVisitor } from "../../holder/capability"
import { issuerGovernanceVisitor } from "../../governance/issuer"


export namespace TestUtil {
  export const CAPABILITY_TEST_CREDENTIAL_TYPE = 'TestCapabilityCredential'

  export type CapabilityCredentialTestParams = TestDocPayload & TestDocExtension

  export type TestDocPayload = {
    description: string
  }

  export type TestDocExtension = {
    info: string
  }

  export type GovCapabilityPresentation = Presentation<EntityIdentity | CapabilityCredential>

  export const CRED_TYPE = 'TestCapabilityBasedCredential1'

  export type TestCredential = Credential<CredentialSubject<
    WrappedDocument<TestDocPayload>, TestDocExtension
  >>

  export type TestClaim = ClaimCredential<ClaimSubject<
    UnsignedCredential<CredentialSubject<
      WrappedDocument<TestDocPayload>, TestDocExtension
    >>
  >>

  export type TestOffer = OfferCredential<OfferSubject<TestCredential, ByCapabilityExtension>>

  export class Wallet extends AgentTestUtil.Wallet {

    static async setup(name: string) {
      const walletUtil = new Wallet(await buildWalletWrapper(nodeCryptoHelper, '11111111',
        { name, alias: name },
        {
          prefix: process.env.DID_PREFIX,
          defaultSchema: 'https://owlmeans.com'
        }
      ), name)

      return walletUtil
    }

    async produceIdentity() {
      return await identityHelper<AgentTestUtil.IdentityFields>(
        this.wallet,
        this.wallet.ssi.buildContext('identity', {
          xsd: 'http://www.w3.org/2001/XMLSchema#',
          firstname: { '@id': 'scm:firstname', '@type': 'xsd:string' },
          lastname: { '@id': 'scm:lastname', '@type': 'xsd:string' }
        })
      ).createIdentity(
        AgentTestUtil.IDENTITY_TYPE,
        {
          firstname: this.name,
          lastname: 'Lastname'
        },
      )
    }

    async requestGovernance() {
      const req = await verifierCredentialHelper(this.wallet).request()
        .build({
          '@type': CREDENTIAL_GOVERNANCE_TYPE,
          source: REGISTRY_TYPE_CAPABILITY
        })

      return await verifierCredentialHelper(this.wallet).request().bundle([req])
    }

    async responseGovernance(request: RequestBundle) {
      const { requests } = await holderCredentialHelper(this.wallet)
        .request().unbundle(request)

      return await holderCredentialHelper<
        {}, {}, CapabilityCredential
      >(this.wallet).response().build(requests, request)
    }

    async claimGovernance(
      rootPresentation: Presentation<EntityIdentity | CapabilityCredential>
    ) {
      const {
        entity: rootEntity,
        credentials
      } = await verifierCredentialHelper(this.wallet).response()
        .verify<CapabilityCredential | EntityIdentity>(rootPresentation)

      const identity = this.wallet.getRegistry(REGISTRY_TYPE_IDENTITIES)
        .getCredential()?.credential as Identity

      return await governanceCredentialHelper(this.wallet).claimGovernance(
        identity,
        {
          root: credentials[0],
          name: 'Organization Governance'
        }
      )
    }

    async selfIssueGovernance(idPresentation: Presentation<EntityIdentity>) {
      const identity = idPresentation.verifiableCredential[0].credentialSubject.data.identity

      const claim = await governanceCredentialHelper(this.wallet).claimGovernance(
        identity, { name: 'Root Governance' }
      )

      const offer = await governanceCredentialHelper(this.wallet).offer(claim)

      const offerBundle = await issuerCredentialHelper(this.wallet)
        .bundle<ClaimCapability, OfferCapability>().build([offer])

      return await holderCredentialHelper<
        CapabilityDocument,
        CapabilityExtension,
        CapabilityCredential,
        OfferCapabilityExtension
      >(
        this.wallet, holderGovernanceVisitor(this.wallet)
      ).bundle().store(offerBundle)
    }

    async claimCapability(
      rootPresentation: Presentation<EntityIdentity | CapabilityCredential>,
      type: string,
      doc: CapabilityCredentialTestParams
    ) {
      const source = this.wallet.getRegistry(REGISTRY_TYPE_IDENTITIES).getCredential()?.credential
      if (!source) {
        throw new Error('No identity to claim for')
      }
      const root = rootPresentation.verifiableCredential.find(
        cap => cap.type.includes(CREDENTIAL_GOVERNANCE_TYPE)
      )

      const claim = await governanceCredentialHelper(this.wallet).claim(
        source,
        {
          root,
          name: 'Test Capability',
          type: [type, CAPABILITY_TEST_CREDENTIAL_TYPE]
        },
        {
          '@type': [type, CAPABILITY_TEST_CREDENTIAL_TYPE],
          subjectSchema: {
            // 'tcap': 'https://example.org/test/capability',
            description: { '@id': 'scm:description', '@type': 'xsd:string' },
            info: { '@id': 'scm:info', '@type': 'xsd:string' },
          },
          subjectProps: {
            payload: {
              description: doc.description
            },
            extension: {
              info: doc.info
            }
          }
        }
      )

      const claimPres = await holderCredentialHelper(this.wallet).bundle().build([claim])
      await holderCredentialHelper(this.wallet).claim({ type }).register(claimPres)

      return claimPres as Presentation<ClaimCapability>
    }

    async signCapability(claimPres: Presentation<ClaimCapability>) {
      const { result, claims, entity }
        = await issuerCredentialHelper(this.wallet)
          .bundle<ClaimCapability, OfferCapability>().unbudle(claimPres)

      const offers = await Promise.all(claims.map(
        async claim => await governanceCredentialHelper(this.wallet).offer(claim)
      ))

      return await issuerCredentialHelper(this.wallet)
        .bundle<ClaimCapability, OfferCapability>().build(offers)
    }

    async storeCapability(offerBundle: OfferBundle<OfferCapability>) {
      const { result } = await holderCredentialHelper<
        CapabilityDocument, CapabilityExtension,
        CapabilityCredential, OfferCapabilityExtension
      >(this.wallet, holderGovernanceVisitor(this.wallet))
        .bundle().unbudle(offerBundle)
      if (!result) {
        throw new Error('Invalid bundle with capability')
      }

      return await holderCredentialHelper<
        CapabilityDocument, CapabilityExtension,
        CapabilityCredential, OfferCapabilityExtension
      >(this.wallet, holderGovernanceVisitor(this.wallet))
        .bundle().store(offerBundle)
    }

    async claimCapabilityCreds(type: string, data: CapabilityCredentialTestParams[]) {
      const claims = await Promise.all(data.map(
        async data => await holderCredentialHelper<
          TestDocPayload,
          TestDocExtension,
          TestCredential
        >(this.wallet)
          .claim({
            type,
            crdContext: {
              description: { '@id': 'scm:description', '@type': 'xsd:string' },
              info: { '@id': 'scm:info', '@type': 'xsd:string' },
            }
          }).build(
            { description: data.description },
            { extension: { info: data.info } }
          )
      ))

      const requestClaim = await holderCredentialHelper(this.wallet)
        .bundle<TestClaim>().build(claims)

      await holderCredentialHelper<TestDocPayload, TestDocExtension, TestCredential>(this.wallet)
        .claim({ type }).register(requestClaim)

      return requestClaim
    }

    async offerCapabilityCreds(claimRequest: ClaimBundle<TestClaim>) {
      const { result, claims } = await issuerCredentialHelper(this.wallet)
        .bundle<TestClaim, TestOffer>().unbudle(claimRequest)

      if (!result) {
        console.log(claimRequest)

        throw new Error('Capability request is broken')
      }

      const offers = await issuerCredentialHelper<
        TestDocPayload,
        TestDocExtension,
        TestCredential,
        ByCapabilityExtension
      >(this.wallet, issuerVisitor(this.wallet)).claim().signClaims(claims)

      return await issuerCredentialHelper(this.wallet)
        .bundle<TestClaim, TestOffer>().build(offers)
    }

    async storeCapabilityCreds(offer: OfferBundle<TestOffer>) {
      const { result } = await holderCredentialHelper<
        TestDocPayload,
        TestDocExtension,
        TestCredential,
        ByCapabilityExtension
      >(
        this.wallet,
        holderCapabilityVisitor<TestDocPayload, TestDocExtension>()(this.wallet)
      ).bundle().unbudle(offer)

      if (!result) {
        // console.log(offer)
        throw new Error('Offer is broken and can\'t be stored')
      }

      await holderCredentialHelper<
        TestDocPayload,
        TestDocExtension,
        TestCredential,
        ByCapabilityExtension
      >(
        this.wallet,
        holderCapabilityVisitor<TestDocPayload, TestDocExtension>()(this.wallet)
      ).bundle().store(offer)
    }

    async provideCredsByCaps(request: RequestBundle) {
      const { requests } = await holderCredentialHelper(this.wallet)
        .request().unbundle(request)

      return await holderCredentialHelper<
        TestDocPayload,
        TestDocExtension,
        TestCredential,
        ByCapabilityExtension
      >(this.wallet, holderCapabilityVisitor<
        TestDocPayload,
        TestDocExtension
      >()(this.wallet))
        .response().build(requests, request)
    }

    async validateResponse<Type extends Credential = TestCredential>(
      response: Presentation<EntityIdentity | Type>
    ) {
      const { result } = await verifierCapabilityHelper<Type>(this.wallet)
        .response().verify(response)

      if (!result) {
        return false
      }

      return true
    }
  }
}