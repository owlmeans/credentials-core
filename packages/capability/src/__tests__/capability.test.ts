require('dotenv').config()

import { TestUtil as Util } from '../debug/utils/wallet'
import {
  capabilityDid,
  entityShape,
  organizationDid,
} from '../debug/utils/shapes'
import {
  credentialShape,
  didShape,
  doubleDidShape,
  keyLessDoubleDidShape,
  presentationShape,
  satelliteShape
} from '@owlmeans/regov-ssi-agent/src/debug/utils/shapes'

import util from 'util'
import { REGISTRY_TYPE_CLAIMS } from '@owlmeans/regov-ssi-core'
import { REGISTRY_TYPE_CAPABILITY } from '..'
util.inspect.defaultOptions.depth = 8


let ctx: {
  alice: Util.Wallet
  bob: Util.Wallet
  charly: Util.Wallet
  dan: Util.Wallet
  fred: Util.Wallet
}

beforeAll(async () => {
  ctx = {
    alice: await Util.Wallet.setup('alice'),
    bob: await Util.Wallet.setup('bob'),
    charly: await Util.Wallet.setup('charly'),
    dan: await Util.Wallet.setup('dan'),
    fred: await Util.Wallet.setup('fred'),
  }

  await ctx.alice.produceIdentity()
  await ctx.bob.produceIdentity()
  await ctx.charly.produceIdentity()
  await ctx.dan.produceIdentity()
  await ctx.fred.produceIdentity()

  const identity = await ctx.charly.provideIdentity()

  await ctx.alice.trustIdentity(identity)
  await ctx.fred.trustIdentity(identity)
  await ctx.bob.trustIdentity(identity)
  await ctx.dan.trustIdentity(identity)

  await ctx.charly.selfIssueGovernance()
})

describe('Capability helpers', () => {
  it('allows to request organization-like capability from an authorithy', async () => {
    const claim = await ctx.fred.claimOrganization('Fredies')
    const offer = await ctx.charly.signCapability(claim)
    await ctx.fred.storeCapability(offer)

    const wraps = await ctx.fred.wallet.getRegistry(REGISTRY_TYPE_CAPABILITY).lookupCredentials(
      Util.ORGANIZATION_CAPABILITY_TYPE
    )

    expect(wraps[0].credential).toMatchSnapshot({
      ...credentialShape,
      credentialSubject: {
        source: {
          ...credentialShape,
          credentialSubject: {
            source: credentialShape,
            sourceDid: didShape
          }
        },
        sourceDid: organizationDid
      }
    })
  })
  // it('allow to self issue gavernance', async () => {
  //   const [governance] = await ctx.charly.selfIssueGovernance(await ctx.charly.provideIdentity())
  //   expect(governance.credential).toMatchSnapshot(govrnanceShape)
  // })

  // it('allow to request governance capability', async () => {
  //   const request = await ctx.bob.requestGovernance()
  //   ctx.gov = await ctx.charly.responseGovernance(request)
  //   expect(ctx.gov).toMatchSnapshot({
  //     ...presentationShape,
  //     verifiableCredential: [
  //       entityShape,
  //       govrnanceShape,
  //       capabilitySatelliteShape
  //     ]
  //   })
  // })

  // it('allow to claim capability', async () => {
  //   if (!ctx.gov) {
  //     throw new Error('No gov capability from previous test')
  //   }
  //   const claim = await ctx.bob.claimCapability(
  //     ctx.gov, Util.CRED_TYPE, {
  //     description: 'Test capability 1',
  //     info: 'Info for capability 1'
  //   })

  //   const offer = await ctx.charly.signCapability(claim)
  //   expect(offer).toMatchSnapshot({
  //     ...presentationShape,
  //     verifiableCredential: [
  //       entityShape,
  //       {
  //         ...credentialShape,
  //         credentialSubject: {
  //           chain: [
  //             didShape,
  //             didShape
  //           ],
  //           did: capabilityDid,
  //           data: {
  //             credential: capabilityShape
  //           }
  //         }
  //       }
  //     ]
  //   })

  //   await ctx.bob.storeCapability(offer)
  // })

  // it('allow to claim credential by capability', async () => {
  //   const claim = await ctx.alice.claimCapabilityCreds(
  //     Util.CRED_TYPE,
  //     [{
  //       description: 'Alice cred',
  //       info: 'Capability based cred for Alice'
  //     }]
  //   )

  //   const offer = await ctx.bob.offerCapabilityCreds(claim)
  //   expect(offer).toMatchSnapshot({
  //     ...presentationShape,
  //     verifiableCredential: [
  //       entityShape,
  //       {
  //         ...credentialShape,
  //         credentialSubject: {
  //           capabilities: [
  //             capabilityShape,
  //             govrnanceShape,
  //           ],
  //           chain: [
  //             capabilityDid,
  //             didShape,
  //             didShape
  //           ],
  //           did: doubleDidShape,
  //           data: {
  //             credential: credentialShape
  //           }
  //         }
  //       }
  //     ]
  //   })

  //   await ctx.alice.storeCapabilityCreds(offer)
  // })

  // it('allow to request capability based credentials', async () => {
  //   const request = await ctx.dan.requestCreds(Util.CRED_TYPE)
  //   const response = await ctx.alice.provideCredsByCaps(request)
  //   expect(response).toMatchSnapshot({
  //     ...presentationShape,
  //     verifiableCredential: [
  //       entityShape,
  //       credentialShape,
  //       {
  //         ...satelliteShape,
  //         credentialSubject: {
  //           data: {
  //             did: keyLessDoubleDidShape,
  //             capabilities: [
  //               capabilityShape
  //             ],
  //             chain: [
  //               capabilityDid,
  //               didShape,
  //               didShape
  //             ],
  //           }
  //         }
  //       }
  //     ]
  //   })
  //   const result = await ctx.dan.validateResponse<Util.TestCredential>(response)
  //   expect(result).toBe(true)
  // })
})