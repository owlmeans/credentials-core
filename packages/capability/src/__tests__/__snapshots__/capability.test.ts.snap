// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Capability helpers allow to request governance capability 1`] = `
Object {
  "@context": Array [
    "https://www.w3.org/2018/credentials/v1",
  ],
  "holder": Object {
    "id": Any<String>,
  },
  "id": Any<String>,
  "proof": Object {
    "challenge": Any<String>,
    "created": Any<String>,
    "domain": Any<String>,
    "jws": Any<String>,
    "proofPurpose": "authentication",
    "type": "EcdsaSecp256k1Signature2019",
    "verificationMethod": Any<String>,
  },
  "type": Array [
    "VerifiablePresentation",
    "CredentialResponse",
  ],
  "verifiableCredential": Array [
    Object {
      "@context": Array [
        "https://www.w3.org/2018/credentials/v1",
        Object {
          "@version": 1.1,
          "data": Object {
            "@context": Object {
              "@version": 1.1,
              "scmdata": "https://example.org/entity/idnetity/data#",
            },
            "@id": "scmdata:id",
            "@type": "scmdata:type",
          },
          "did": Object {
            "@id": "scm:did",
            "@type": "@json",
          },
          "identity": Object {
            "@id": "scm:identity",
            "@type": "@json",
          },
          "scm": "https://example.org/entity/idnetity#",
        },
      ],
      "credentialSubject": Object {
        "data": Object {
          "@type": "EntityIdentity",
          "identity": Object {
            "@context": Array [
              "https://www.w3.org/2018/credentials/v1",
              Object {
                "@version": 1.1,
                "data": Object {
                  "@context": Object {
                    "@version": 1.1,
                    "scmdata": "https://example.org/identity/data#",
                  },
                  "@id": "scmdata:id",
                  "@type": "scmdata:type",
                },
                "firstname": Object {
                  "@id": "scm:firstname",
                  "@type": "xsd:string",
                },
                "lastname": Object {
                  "@id": "scm:lastname",
                  "@type": "xsd:string",
                },
                "scm": "https://example.org/identity#",
                "xsd": "http://www.w3.org/2001/XMLSchema#",
              },
            ],
            "credentialSubject": Object {
              "data": Object {
                "@type": "TestUtilIdentity",
                "firstname": "charly",
                "lastname": "Lastname",
              },
            },
            "holder": Object {
              "id": Any<String>,
            },
            "id": Any<String>,
            "issuanceDate": Any<String>,
            "issuer": Any<String>,
            "proof": Object {
              "created": Any<String>,
              "jws": Any<String>,
              "proofPurpose": "assertionMethod",
              "type": "EcdsaSecp256k1Signature2019",
              "verificationMethod": Any<String>,
            },
            "type": Array [
              "VerifiableCredential",
              "TestUtilIdentity",
            ],
          },
        },
        "did": Object {
          "@context": Array [
            "https://w3id.org/did/v1",
            "https://w3id.org/security/v2",
          ],
          "assertionMethod": Array [
            Any<String>,
          ],
          "authentication": Array [
            Any<String>,
          ],
          "capabilityDelegation": Array [
            Any<String>,
          ],
          "capabilityInvocation": Array [
            Any<String>,
          ],
          "id": Any<String>,
          "keyAgreement": Array [
            Any<String>,
          ],
          "proof": Object {
            "created": Any<String>,
            "jws": Any<String>,
            "proofPurpose": "verificationMethod",
            "type": "EcdsaSecp256k1Signature2019",
            "verificationMethod": Any<String>,
          },
          "verificationMethod": Array [
            Object {
              "controller": Any<String>,
              "id": Any<String>,
              "nonce": Any<String>,
              "publicKeyBase58": Any<String>,
              "type": "EcdsaSecp256k1VerificationKey2019",
            },
          ],
        },
      },
      "holder": Object {
        "id": Any<String>,
      },
      "id": Any<String>,
      "issuanceDate": Any<String>,
      "issuer": Any<String>,
      "proof": Object {
        "created": Any<String>,
        "jws": Any<String>,
        "proofPurpose": "assertionMethod",
        "type": "EcdsaSecp256k1Signature2019",
        "verificationMethod": Any<String>,
      },
      "type": Array [
        "VerifiableCredential",
        "EntityIdentity",
      ],
    },
    Object {
      "@context": Array [
        "https://www.w3.org/2018/credentials/v1",
        Object {
          "@version": 1.1,
          "credentialProps": Object {
            "@id": "scm:credentialProps",
            "@type": "@json",
          },
          "credentialSchema": Object {
            "@id": "scm:credentialSchema",
            "@type": "@json",
          },
          "data": Object {
            "@context": Object {
              "@version": 1.1,
              "scmdata": "https://example.org/governance/capability/data#",
            },
            "@id": "scmdata:id",
            "@type": "scmdata:type",
          },
          "name": Object {
            "@id": "scm:name",
            "@type": "xsd:string",
          },
          "root": Object {
            "@id": "scm:root",
            "@type": "xsd:root",
          },
          "scm": "https://example.org/governance/capability#",
          "selfIssuing": Object {
            "@id": "scm:selfIssuing",
            "@type": "xsd:boolean",
          },
          "source": Object {
            "@id": "scm:source",
            "@type": "@json",
          },
          "subjectProps": Object {
            "@id": "scm:subjectProps",
            "@type": "@json",
          },
          "subjectSchema": Object {
            "@id": "scm:subjectSchema",
            "@type": "@json",
          },
        },
      ],
      "credentialSubject": Object {
        "data": Object {
          "@type": Array [
            "GovernanceCapability",
          ],
        },
        "name": "Governance Claim",
        "source": Object {
          "@context": Array [
            "https://w3id.org/did/v1",
            "https://w3id.org/security/v2",
          ],
          "assertionMethod": Array [
            Any<String>,
          ],
          "authentication": Array [
            Any<String>,
          ],
          "capabilityDelegation": Array [
            Any<String>,
          ],
          "capabilityInvocation": Array [
            Any<String>,
          ],
          "id": Any<String>,
          "keyAgreement": Array [
            Any<String>,
          ],
          "proof": Object {
            "created": Any<String>,
            "jws": Any<String>,
            "proofPurpose": "verificationMethod",
            "type": "EcdsaSecp256k1Signature2019",
            "verificationMethod": Any<String>,
          },
          "verificationMethod": Array [
            Object {
              "controller": Any<String>,
              "id": Any<String>,
              "nonce": Any<String>,
              "publicKeyBase58": Any<String>,
              "type": "EcdsaSecp256k1VerificationKey2019",
            },
          ],
        },
      },
      "holder": Object {
        "id": Any<String>,
      },
      "id": Any<String>,
      "issuanceDate": Any<String>,
      "issuer": Any<String>,
      "proof": Object {
        "created": Any<String>,
        "jws": Any<String>,
        "proofPurpose": "assertionMethod",
        "type": "EcdsaSecp256k1Signature2019",
        "verificationMethod": Any<String>,
      },
      "type": Array [
        "VerifiableCredential",
        "CredentialCapability",
        "GovernanceCapability",
      ],
    },
    Object {
      "@context": Array [
        "https://www.w3.org/2018/credentials/v1",
        Object {
          "@version": 1.1,
          "data": Object {
            "@context": Object {
              "@version": 1.1,
              "scmdata": "https://example.org/credential/satellite/data#",
            },
            "@id": "scmdata:id",
            "@type": "scmdata:type",
          },
          "did": Object {
            "@id": "scm:did",
            "@type": "@json",
          },
          "scm": "https://example.org/credential/satellite#",
        },
      ],
      "credentialSubject": Object {
        "data": Object {
          "@type": "CredentialSatellit",
          "did": Object {
            "@context": Array [
              "https://w3id.org/did/v1",
              "https://w3id.org/security/v2",
            ],
            "assertionMethod": Array [
              Any<String>,
            ],
            "authentication": Array [
              Any<String>,
            ],
            "capabilityDelegation": Array [
              Any<String>,
            ],
            "capabilityInvocation": Array [
              Any<String>,
            ],
            "id": Any<String>,
            "proof": Object {
              "created": Any<String>,
              "jws": Any<String>,
              "proofPurpose": "verificationMethod",
              "type": "EcdsaSecp256k1Signature2019",
              "verificationMethod": Any<String>,
            },
            "verificationMethod": Array [
              Object {
                "controller": Any<String>,
                "id": Any<String>,
                "nonce": Any<String>,
                "publicKeyBase58": Any<String>,
                "type": "EcdsaSecp256k1VerificationKey2019",
              },
            ],
          },
        },
      },
      "holder": Object {
        "id": Any<String>,
      },
      "id": Any<String>,
      "issuanceDate": Any<String>,
      "issuer": Any<String>,
      "proof": Object {
        "created": Any<String>,
        "jws": Any<String>,
        "proofPurpose": "assertionMethod",
        "type": "EcdsaSecp256k1Signature2019",
        "verificationMethod": Any<String>,
      },
      "type": Array [
        "VerifiableCredential",
        "CredentialSatellit",
      ],
    },
  ],
}
`;

exports[`Capability helpers allow to self issue gavernance 1`] = `
Object {
  "@context": Array [
    "https://www.w3.org/2018/credentials/v1",
    Object {
      "@version": 1.1,
      "credentialProps": Object {
        "@id": "scm:credentialProps",
        "@type": "@json",
      },
      "credentialSchema": Object {
        "@id": "scm:credentialSchema",
        "@type": "@json",
      },
      "data": Object {
        "@context": Object {
          "@version": 1.1,
          "scmdata": "https://example.org/governance/capability/data#",
        },
        "@id": "scmdata:id",
        "@type": "scmdata:type",
      },
      "name": Object {
        "@id": "scm:name",
        "@type": "xsd:string",
      },
      "root": Object {
        "@id": "scm:root",
        "@type": "xsd:root",
      },
      "scm": "https://example.org/governance/capability#",
      "selfIssuing": Object {
        "@id": "scm:selfIssuing",
        "@type": "xsd:boolean",
      },
      "source": Object {
        "@id": "scm:source",
        "@type": "@json",
      },
      "subjectProps": Object {
        "@id": "scm:subjectProps",
        "@type": "@json",
      },
      "subjectSchema": Object {
        "@id": "scm:subjectSchema",
        "@type": "@json",
      },
    },
  ],
  "credentialSubject": Object {
    "data": Object {
      "@type": Array [
        "GovernanceCapability",
      ],
    },
    "name": "Governance Claim",
    "source": Object {
      "@context": Array [
        "https://w3id.org/did/v1",
        "https://w3id.org/security/v2",
      ],
      "assertionMethod": Array [
        Any<String>,
      ],
      "authentication": Array [
        Any<String>,
      ],
      "capabilityDelegation": Array [
        Any<String>,
      ],
      "capabilityInvocation": Array [
        Any<String>,
      ],
      "id": Any<String>,
      "keyAgreement": Array [
        Any<String>,
      ],
      "proof": Object {
        "created": Any<String>,
        "jws": Any<String>,
        "proofPurpose": "verificationMethod",
        "type": "EcdsaSecp256k1Signature2019",
        "verificationMethod": Any<String>,
      },
      "verificationMethod": Array [
        Object {
          "controller": Any<String>,
          "id": Any<String>,
          "nonce": Any<String>,
          "publicKeyBase58": Any<String>,
          "type": "EcdsaSecp256k1VerificationKey2019",
        },
      ],
    },
  },
  "holder": Object {
    "id": Any<String>,
  },
  "id": Any<String>,
  "issuanceDate": Any<String>,
  "issuer": Any<String>,
  "proof": Object {
    "created": Any<String>,
    "jws": Any<String>,
    "proofPurpose": "assertionMethod",
    "type": "EcdsaSecp256k1Signature2019",
    "verificationMethod": Any<String>,
  },
  "type": Array [
    "VerifiableCredential",
    "CredentialCapability",
    "GovernanceCapability",
  ],
}
`;
